File: longest-univalue-path/longest-univalue-path.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public class Node{
        public int val;
        public int count;
        public Node(int val, int count) {
            this.val = val;
            this.count = count;
        }
    }
    private int len;
    public int LongestUnivaluePath(TreeNode root) {
        Helper(root);
        return Math.Max(0, len - 1);
    }
    private Node Helper(TreeNode node) {
        if(node == null) {
            return null;
        }
        if(node.left == null && node.right == null) {
            return new Node(node.val, 1);
        }
        Node left = Helper(node.left);
        Node right = Helper(node.right);
        int count = 1;
        int max = 0;
        if(left != null && node.val == left.val) {
            count += left.count;
            max = Math.Max(left.count, max);
        }
        if(right != null && node.val == right.val) {
            count += right.count;
            max = Math.Max(right.count, max);
        }
        len = Math.Max(count, len);
        return new Node(node.val, max + 1);
    }
}

File: longest-univalue-path/longest-univalue-path.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int len;
    public int longestUnivaluePath(TreeNode root) {
        if(root == null) return 0;
        len = -1001;
        helper(root);
        return len == -1001 ? 0 : len - 1;
    }
    private int helper(TreeNode node) {
        if(node == null) return 0;
        int left = helper(node.left);
        int right = helper(node.right);
        int count = 1;
        if(node.left != null && node.right != null && node.left.val == node.val && node.right.val == node.val) {
            len = Math.max(len, left + right + 1);
            count += Math.max(left, right);
        }
        else if(node.left != null && node.left.val == node.val) {
            len = Math.max(len, left + 1);
            count += left;
        }
        else if(node.right != null && node.right.val == node.val) {
            len = Math.max(len, right + 1);
            count += right;
        }
        return count;
    }
}